shell-completion/zsh/_systemctl.in

```diff
@@ -424,20 +424,13 @@ done
 (( $+functions[_systemctl_caching_policy] )) ||
     _systemctl_caching_policy()
 {
-    local _sysunits
-    local -a oldcache
-
     # rebuild if cache is more than a day old
-    oldcache=( "$1"(mh+1) )
-    (( $#oldcache )) && return 0
-
-    _sysunits=(${${(f)"$(__systemctl --all)"}%% *})
+    [[ -n $1(#qNmd+1) ]] && return 0
 
-    if (( $#_sysunits )); then
-        for unit in $_sysunits; do
-            [[ "$unit" -nt "$1" ]] && return 0
-        done
-    fi
+    local pathkind=systemd-search-${1##*--}-path
+    for dir in ${(s-:-)^$(_call_program $pathkind systemd-path $pathkind)}; do
+        [[ $dir -nt $1 ]] && return 0
+    done
```

Comment:
I'm not sure what is going on, but if I do `systemctl enable <TAB>`, it feels like I always have to wait the same amount of time. If I call `_systemctl_caching_policy; echo $?` I always get 0.

Reply:
The policy takes an argument to the cache identifier (the filename under the cache-path) in $1. The return code if that cache file doesn't exist also isn't very relevant, since it will be rebuilt based on the return value of _retrieve_cache in that case. Thanks for asking though, b/c I see I made a typo in pathkind — it should be systemd-search-system-unit, not systemd-search-system-path...

Anyway your trouble however is that currently there is no cache used in that operation. The completions here try to be specific in only offering "disabled" units, and the currently implemented caches which hold units names just aren't useful for that. So, it instead parses list-unit-files again — twice, in fact, to first get the list of disabled units and again to get the list of unit files templates. I don't know if there's a reasonable way to cache the state of units from the output of list-unit-files (i.e. does the dir mtime check still work for validity?), but templates are just filenames, so we should be able to cut that time roughly in half by glob matching them.

Reply:
Latest commit may help speed up enable.